### ISSUE TITLE
Confusion about masking in transformer

### ISSUE THREAD

**Issue by Yupei-Du** [bentrevett/pytorch-seq2seq#119]
Thank you for your tutorial! However, I have some confusion about masking.  

In ```make_src_mask```, ```src_mask = (src != self.src_pad_idx).unsqueeze(1).unsqueeze(2)```. 
w.l.o.g. let the ```batch_size = 1```: for example, a batch consists of only a sentence ```I love NLP <pad>```. After broadcasting in ```MultiHeadAttention```, the mask would be ```[[1110],[1110],[1110],[1110]]```. However, since the 4th row corresponds to the token ```<pad>```, the mask should be ```[0000]```. 

Did I make any mistake or have I made my point clear? Thank you a lot in advance!

**Comment by bentrevett** [comment]
Not sure I fully understand, here's a quick code walk-through with a batch size of 1.

We can see the mask is correctly created, broadcast, and applied to the `energy` tensor.

```python
import torch
torch.manual_seed(0)

pad_idx = 0

sequence = torch.LongTensor([[1, 2, 3, 0]]) #[batch, seq length], where 0 = padding

print(sequence.shape) #torch.Size([1,4]) [batch, seq length]

mask = (sequence != pad_idx).unsqueeze(1).unsqueeze(2)

print(mask) #tensor([[[[ True,  True,  True, False]]]]) where False corresponds to padding
print(mask.shape) #torch.Size([1,1,1,4]), expanded to required size for multi-headed attenton

#in the multi-headed attention layer
#create random energy tensor

energy = torch.randn(1, 2, 3, 4) #[batch, n heads, query len, key len]

print(energy)
#tensor([[[[-1.1258, -1.1524, -0.2506, -0.4339],
#          [ 0.8487,  0.6920, -0.3160, -2.1152],
#          [ 0.4681, -0.1577,  1.4437,  0.2660]],
#
#         [[ 0.1665,  0.8744, -0.1435, -0.1116],
#          [ 0.9318,  1.2590,  2.0050,  0.0537],
#          [ 0.6181, -0.4128, -0.8411, -2.3160]]]])
#two heads, so two 3x4 (query len x key len) blocks

energy = energy.masked_fill(mask == 0, -1e10)

print(energy)
#tensor([[[[-1.1258e+00, -1.1524e+00, -2.5058e-01, -1.0000e+10],
#          [ 8.4871e-01,  6.9201e-01, -3.1601e-01, -1.0000e+10],
#          [ 4.6810e-01, -1.5771e-01,  1.4437e+00, -1.0000e+10]],
#
#         [[ 1.6646e-01,  8.7438e-01, -1.4347e-01, -1.0000e+10],
#          [ 9.3183e-01,  1.2590e+00,  2.0050e+00, -1.0000e+10],
#          [ 6.1806e-01, -4.1280e-01, -8.4106e-01, -1.0000e+10]]]])
#correctly masks out the last sequence element among the key dim

#prints last key dim per query in the first head, which should be all masked
print(energy[0,0,:,-1]) #tensor([-1.0000e+10, -1.0000e+10, -1.0000e+10])
#which they are

print(energy[0,1,:,-1]) #tensor([-1.0000e+10, -1.0000e+10, -1.0000e+10])
#same for the second head
```

Let me know if I've made a mistake somewhere or there's anything that needs clarifying.

**Comment by Yupei-Du** [comment]
Thank you so much for your quick response! Yes, from your explanation I have understood the rationale behind this masking strategy.

**Comment by Daniel63656** [comment]
I have the same question but reading your explanation my doubts are not resolved. This basically doesn't take into account that padding can also exist in the **query** as well. If we look at your example and assume we are in the encoder-self-attention (query=key), than in fact the padding mask should look like this:

T   T   T   F
T   T   T   F
T   T   T   F
F   F   F   F

Don't know how you would call such a matrix or why nobody seems the necessity to implement it this way. Am I overseeing something here?
